### PR TITLE
DRY up AWS resource implementation and test backend implementations

### PR DESCRIPTION
**Pull request by clintoncwolfe**
This PR provides some standards and facilities for accelerating creation of new AWS resources.

- Provides a standard constructor and resource parameter validation and unpacking mechanism.
- Implements `exists?`  (often overlooked)
- DRYs up the creation of an AWS API mocking facility for unit testing
- Moves library support code to a clear location
- Separates the AWS connection code from the shimming mechanism (partially implements #95)

No docs changes.  Unit and integration tests still pass.

### CODE REVIEW

**Review by miah — APPROVED** [review]

**Comment by aaronlippold** [comment]
Is this good to go? Merge ready?

**Comment by clintoncwolfe** [comment]
We have another internal review yet to pass, then I think it will be ready for merge.

File: lib/inspec/aws/connection.rb [hunk 1]
```diff
@@ -1,8 +1,13 @@
 # author: Christoph Hartmann
 
+# This class exists so that we can intercept AWS API connection setup
+# and have an opportunity to provide credentials from another mechanism
+# (such as a train transport URI) in the future.
+#
+# We commit to always supporting the standard AWS environment variables.
+
```
Comment (jerryaldrichiii) [inline lib/inspec/aws/connection.rb]: Empty line alert :smile:

**Review by jerryaldrichiii — APPROVED** [review]
Looks great! I didn't see anything after our internal review save for a measly empty line.

**Review by arlimus — APPROVED** [review]
So DRYyyyy - sweet work + reviews thank you all!!

**Pull request merged**